src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+        })?;
+        fs::remove_dir_all(&self.wal_path).with_context(|| {
+            format!("failed to remove wal directory {}", self.wal_path.display())
+        })?;
+
+        Ok(())
+    }
+
+    /// Checks if there are finished WAL segments and if so, forms them into batches.
+    fn consume_wal(&mut self) -> Result<(), anyhow::Error> {
+        let finished_segments = self.find_finished_wal_segments()?;
+
+        for segment in finished_segments {
+            let batch = Batch::create(&segment, &self.trace_path)?;
+            self.batches.push(batch);
+            fs::remove_file(&segment).with_context(|| {
+                format!(
+                    "failed to remove consumed wal segment {}",
+                    segment.display()
+                )
+            })?;
+        }
+
+        Ok(())
+    }
+
+    fn find_finished_wal_segments(&self) -> Result<Vec<PathBuf>, anyhow::Error> {
+        lazy_static! {
+            static ref FINISHED_WAL_SEGMENT_REGEX: Regex =
+                Regex::new("^log-[0-9]+-final$").unwrap();
+        }
+
+        let mut segments = read_dir_regex(&self.wal_path, &FINISHED_WAL_SEGMENT_REGEX)?;
+        segments.sort_by_key(|segment| {
+            segment
+                .to_str()
+                .unwrap()
+                .split('-')
+                .nth(1)
+                .unwrap()
+                .parse::<usize>()
+                .unwrap()
+        });
+
+        Ok(segments)
+    }
+
+    fn find_unfinished_wal_segment(&self) -> Result<PathBuf, anyhow::Error> {
+        lazy_static! {
+            static ref UNFINISHED_WAL_SEGMENT_REGEX: Regex = Regex::new("^log-[0-9]+$").unwrap();
+        }
+
+        let mut segments = read_dir_regex(&self.wal_path, &UNFINISHED_WAL_SEGMENT_REGEX)?;
+        if segments.len() > 1 {
+            bail!(
+                "Expected only a single unfinished wal segment at {}. Found {}",
+                self.wal_path.display(),
+                segments.len()
+            );
+        }
+
+        if segments.len() == 0 {
+            bail!(
+                "Expected at least a single unfinished wal segment at {}. Found none.",
+                self.wal_path.display()
+            );
+        }
+
+        Ok(segments.pop().unwrap())
+    }
+
+    fn find_batches(&self) -> Result<Vec<Batch>, anyhow::Error> {
+        lazy_static! {
+            static ref BATCH_REGEX: Regex = Regex::new("^batch-[0-9]+-[0-9]+$").unwrap();
+        }
+
+        let batches = read_dir_regex(&self.trace_path, &BATCH_REGEX)?;
+
+        let mut batches: Vec<Batch> = batches
+            .into_iter()
+            .map(Batch::reinit)
+            .collect::<Result<_, _>>()
+            .unwrap();
+        batches.sort_by_key(|batch| batch.lower);
+        Ok(batches)
+    }
+
+    /// Try to compact all of the batches we know about into a single batch from
+    /// [compaction_frontier, upper)
+    ///
+    /// TODO: the approach to compacting is likely very suboptimal.
```

Comment:
It is! :D

If you apply the same test (or any test, really) to groups that have the same order of magnitude length, that works out much better. This will start to choke when there is one batch with 100GB of data and 9 containing 1MB, with the method called each second.